File: apps/mobile/modules/nemu-aidoku/ios/NemuNativeHttpRedirectPolicy.swift
import Foundation

/// Per-request redirect policy for credential-bearing native HTTP calls.
/// General Aidoku source traffic may intentionally use HTTP, so callers opt in
/// only when every hop must remain authenticated HTTPS (for example PKCE token
/// exchange). URLSession asks this policy before it follows each redirect.
enum NemuNativeHttpRedirectPolicy {
  static let blockedMessage =
    "Native source networking blocked a redirect that did not remain HTTPS."

  static func allows(_ url: URL?, requireHttps: Bool) -> Bool {
    guard requireHttps else { return true }
    return url?.scheme?.lowercased() == "https"
  }
}

File: apps/mobile/modules/nemu-aidoku/runtime/kotlin/NemuNativeHttpRedirectPolicy.kt
package pm.nemu.mobile.aidoku

import java.io.IOException
import okhttp3.Interceptor
import okhttp3.Response

/** Marker copied by OkHttp onto follow-up requests for credential-bearing calls. */
internal object NemuHttpsOnlyRequestPolicy {
  const val BLOCKED_MESSAGE =
    "Native source networking blocked a redirect that did not remain HTTPS."

  fun allows(url: okhttp3.HttpUrl): Boolean = url.isHttps
}

/**
 * Network interceptors run for every redirect hop. Rejecting before
 * [Interceptor.Chain.proceed] ensures an authorization code, PKCE verifier, or
 * other request body is never written to a cleartext redirected connection.
 */
internal class NemuHttpsOnlyRedirectInterceptor : Interceptor {
  override fun intercept(chain: Interceptor.Chain): Response {
    val request = chain.request()
    if (
      request.tag(NemuHttpsOnlyRequestPolicy::class.java) != null &&
      !NemuHttpsOnlyRequestPolicy.allows(request.url)
    ) {
      throw IOException(NemuHttpsOnlyRequestPolicy.BLOCKED_MESSAGE)
    }
    return chain.proceed(request)
  }
}
